### PR TITLE
Added perf and perf_minopts scenarios for legacy backend

### PR DESCRIPTION
**Pull request by rartemev**

### CODE REVIEW

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by jashook** [comment]
@rartemev have you run run-xunit-perf.cmd with the changes made to verify it runs correctly with legacy backend?

File: netci.groovy [hunk 1]
```diff
@@ -1396,7 +1396,7 @@ def static addTriggers(def job, def branch, def isPR, def architecture, def os,
                     }
                     break
                 default:
-                    println("Unknown scenario: ${os} ${arch} ${jit} ${scenario}");
+                    println("Unknown scenario: ${os} ${architecture} ${jit} ${scenario}");
```
Comment (jashook) [inline netci.groovy]: Why is this change needed?

File: perf.groovy [hunk 2]
```diff
@@ -106,6 +107,10 @@ def static getOSGroup(def os) {
                         batchFile("tests\\runtest.cmd ${configuration} ${architecture} GenerateLayoutOnly")
 
                         // Run with just stopwatch: Profile=Off
+                        if (testEnv != '')
+                        {
+                            batchFile("set TEST_ENV=${testEnv}")
```
Comment (jashook) [inline perf.groovy]: Unfortunately this will have unintended consequences. This creates a batch file with one line: 

`set TEST_ENV=...`

At the end of the execution the environment is lost and TEST_ENV will be unset for the next batch script called.

File: tests/x86/legacyjit_x86__minopts_testenv.cmd [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+@REM  This script provides x86 Legacy JIT with MinOpts test environment settings
+@REM
+@REM -------------------------------------------------------------------------
+
+set COMPLUS_AltJit=*
+set COMPLUS_AltJitNgen=*
+set COMPLUS_AltJitName=legacyjit.dll
+set COMPLUS_NoGuiOnAssert=1
+set COMPLUS_AltJitAssertOnNYI=1
+set COMPlus_JITMinOpts=1
```
Comment (jashook) [inline tests/x86/legacyjit_x86__minopts_testenv.cmd]: I think the better option instead of creating many different test environments with different stress scenarios is to  load a base testenv then add stress modes. It is unclear to me how our performance infrastructure is extensible to stress scenarios like min opts as it is today.

**Comment by jashook** [comment]
I think it is also worth thinking about whether we want to add an x86 folder or not. Currently we have an arm and arm64 folder which hold architecture specific testing information under /tests. Adding an x86 folder to hold test env batch files breaks our current usage for these architecture specific folders. It may be for the better or may be confusing I do not know entirely.

@RussKeldorph opinions

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by RussKeldorph** [comment]
> I think it is also worth thinking about whether we want to add an x86 folder or not. C

I don't have a strong opinion, but I wouldn't be inclined to make that change with this PR.

I agree we should make sure the scenario is completely working (all the benchmarks run as expected) before worrying about the infrastructure (this change), though.

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@jashook @RussKeldorph please review one more time

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
>have you run run-xunit-perf.cmd with the changes made to verify it runs correctly with legacy backend?

@jashook Yes, I have. With that #13132 fix everything runs correctly.

File: perf.groovy [hunk 4]
```diff
@@ -106,12 +113,12 @@ def static getOSGroup(def os) {
...
-                        batchFile("tests\\scripts\\run-xunit-perf.cmd -arch ${arch} -configuration ${configuration} -testBinLoc bin\\tests\\${os}.${architecture}.${configuration}\\Jit\\Performance\\CodeQuality -generateBenchviewData \"%WORKSPACE%\\Microsoft.Benchview.JSONFormat\\tools\" ${uploadString} -runtype ${runType} -stabilityPrefix \"START \"CORECLR_PERF_RUN\" /B /WAIT /HIGH /AFFINITY 0x2\"")
+                        batchFile("${precommand} \n tests\\scripts\\run-xunit-perf.cmd -arch ${arch} -configuration ${configuration} -testBinLoc bin\\tests\\${os}.${architecture}.${configuration}\\performance\\perflab\\Perflab -library -generateBenchviewData \"%WORKSPACE%\\Microsoft.Benchview.JSONFormat\\tools\" ${uploadString} -runtype ${runType} -stabilityPrefix \"START \"CORECLR_PERF_RUN\" /B /WAIT /HIGH /AFFINITY 0x2\"")
```
Comment (RussKeldorph) [inline perf.groovy]: Are you using this mechanism because the `testEnv` functionality is broken?  I think my preference would be to fix testEnv and then create .cmd files for every combination of variables that needs to be set by this .groovy file.  If that is too much work, I'd like to understand why.

If you do need to use `precommand`, were you able to verify it works, like for example testing in the dev/unix_test_workflow branch CI?

Comment (rartemev) [inline perf.groovy]: Actually I don't know whether it is broken, but I figured out that tests\\scripts\\run-xunit-perf.cmd expects TEST_ENV var to be set to .cmd script that sets the test environment.

At the beginning I was using 2 different .cmd files for each env (lb, lb+minopt) but @jashook said it would be better to set minopt option inside the groove script instead (see comment to tests/x86/legacyjit_x86__minopts_testenv.cmd above)

**Comment by jashook** [comment]
After some thought I realized this change may still upload information from the new jobs. However the change does not add any new configuration options so it will upload the legacy backend and legacy backend min opts data as if it was ryujit.

**Comment by adiaaida** [comment]
To what Jarret mentioned: in the past, we had jit32 listed as a separate architecture in benchview, so when we pass submission.py, we'd list x86jit32 as the architecture. You could do something similar for legacy backend (x86legacy, maybe?). For min opts on the throughput runs, we are just passing that as a separate config key value pair, so you'd do

`--config OptLevel min_opt`

for min opts and full_opt for regular. If we wanted to, we could do something like that for legacy vs default here. So:

```
--config Jit legacy
--config Jit ryujit
```

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by jashook** [comment]
@dotnet-bot help

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by jorive** [comment]
@rartemev Is this change something we want to also measure on the JitBench scenario too?

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
Added legacy backend into every perf scenario

@jashook @adiaaida @jorive @RussKeldorph PTAL

File: perf.groovy [hunk 5]
```diff
@@ -516,88 +552,116 @@ parallel(
...
+                            "set \"BENCHVIEWNAME=${benchViewName}\"\n" +
+                            "set \"BENCHVIEWNAME=%BENCHVIEWNAME:\"=%\"\n" +
+                            "py \"%WORKSPACE%\\Microsoft.BenchView.JSONFormat\\tools\\submission-metadata.py\" --name \"%BENCHVIEWNAME%\" --user \"dotnet-bot@microsoft.com\"\n" +
+                            "py \"%WORKSPACE%\\Microsoft.BenchView.JSONFormat\\tools\\build.py\" git --branch %GIT_BRANCH_WITHOUT_ORIGIN% --type ${runType}")
+                            batchFile("py \"%WORKSPACE%\\Microsoft.BenchView.JSONFormat\\tools\\machinedata.py\"")
+                            batchFile("set __TestIntermediateDir=int&&build.cmd ${configuration} ${architecture}")
+
+                            batchFile("tests\\runtest.cmd ${configuration} ${architecture} GenerateLayoutOnly")
+
+                            if (arch == 'x64') {
```
Comment (jorive) [inline perf.groovy]: I believe ILLink should only run under ryujit && full_opt.
@swaroop-sridhar Could you please take a look?

File: perf.groovy [hunk 6]
```diff
@@ -30,120 +30,142 @@ def static getOSGroup(def os) {
     ['Windows_NT'].each { os ->
         ['x64', 'x86'].each { arch ->
             [true, false].each { isSmoketest ->
-                def architecture = arch
-                def jobName = isSmoketest ? "perf_perflab_${os}_${arch}_smoketest" : "perf_perflab_${os}_${arch}"
+                ['ryujit', 'legacy'].each { jit ->
```
Comment (jashook) [inline perf.groovy]: Nit: legacy_backend or lb instead of legacy.

File: perf.groovy (hunk 5 above)
Comment (swaroop-sridhar) [inline perf.groovy]: Sure, I'm OK if we want to change the ILLink leg to (ryujit && full_opt).

File: tests/scripts/run-throughput-perf.py [hunk 7]
```diff
@@ -290,6 +295,11 @@ def runIterations(dll_name, dll_path, iterations, crossgen_path, jit_path, assem
     if opt_level == 'min_opt':
         my_env['COMPlus_JITMinOpts'] = '1'
 
+    if jit_name == 'legacy':
+        my_env['COMPLUS_AltJit'] = '*'
```
Comment (adiaaida) [inline tests/scripts/run-throughput-perf.py]: The way we did this in the past was to give crossgen a different JitPath for the legacy backend run. You may still need to pass AltJit/AltJitNgen, etc. but we had working runs when passing legacyjit.dll to crossgen as the jit. Have you tested this to make sure the timing data looks about right?

File: tests/scripts/run-throughput-perf.py [hunk 8]
```diff
@@ -417,6 +427,9 @@ def main(args):
                 "--config",
                 "OptLevel",
                 opt_level,
+                "--config",
```
Comment (adiaaida) [inline tests/scripts/run-throughput-perf.py]: As Jarret mentions, we used to have jobs for jit32, which we listed as a different architecture to match outerloop tp jobs (it was x86jit32). So we may want to copy that behavior here.

File: tests/scripts/run-throughput-perf.py (hunk 7 above)
Comment (rartemev) [inline tests/scripts/run-throughput-perf.py]: Could you please specify what is "right"?
I have tested and results were pretty different for "COMPlus_AltJitName=", "COMPlus_AltJitName=legacyjit.dll" and "COMPlus_AltJitName=legacyjit.dll && COMPlus_JITMinOpts=1" so I think it is the correct set of EnvVars.

Comment (adiaaida) [inline tests/scripts/run-throughput-perf.py]: That's all that I wanted to confirm. Using legacyjit.dll should produce faster throughput than x86 ryujit.

File: tests/scripts/run-throughput-perf.py (hunk 8 above)
Comment (rartemev) [inline tests/scripts/run-throughput-perf.py]: Changed 'legacy' to 'legacy_backend'

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

File: tests/scripts/run-throughput-perf.py (hunk 7 above)
Comment (BruceForstall) [inline tests/scripts/run-throughput-perf.py]: > Using legacyjit.dll should produce faster throughput than x86 ryujit.

Not necessarily!

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

File: tests/scripts/run-throughput-perf.py (hunk 7 above)
Comment (rartemev) [inline tests/scripts/run-throughput-perf.py]: According to my results on PerfLab legacy was the fastest one, then ryujit, and legacy+minopt is the slowest.

**Comment by rartemev** [comment]
@dotnet-bot test CI Please

**Comment by DrewScoggins** [comment]
test ci please

**Comment by rartemev** [comment]
test ci please

**Comment by rartemev** [comment]
test ci please

**Comment by rartemev** [comment]
test ci please

**Comment by rartemev** [comment]
@dotnet/jit-contrib PTAL

File: tests/scripts/run-throughput-perf.py [hunk 9]
```diff
@@ -165,6 +168,7 @@ def validate_arg(arg, check):
     valid_run_types = ['rolling', 'private']
     valid_os = ['Windows_NT', 'Ubuntu14.04']
     valid_opt_levels = ['full_opt', 'min_opt']
+    valid_jit_names = ['ryujit', 'legacy_backend']
```
Comment (adiaaida) [inline tests/scripts/run-throughput-perf.py]: Since legacy_backend is only valid for x86, I'd like to see this changed to resemble the valid_arches list so:

`valid_jit_names = {'x64': ['ryujit'], 'x86': ['ryujit', 'legacy_backend']}`

File: tests/scripts/run-throughput-perf.py [hunk 10]
```diff
@@ -417,6 +430,9 @@ def main(args):
                 "--config",
                 "OptLevel",
                 opt_level,
+                "--config",
+                "JitName",
+                jit_name,
```
Comment (adiaaida) [inline tests/scripts/run-throughput-perf.py]: I've talked to @RussKeldorph about this, and I'm actually pretty ok with the JitName config option.

File: tests/scripts/run-throughput-perf.py (hunk 9 above)
Comment (rartemev) [inline tests/scripts/run-throughput-perf.py]: Fixed

**Comment by rartemev** [comment]
test ci please

**Comment by jashook** [comment]
Looks good to me and spoke with @DrewScoggins and throughput is not an issue due to more hardware.

**Comment by rartemev** [comment]
@dotnet-bot test CI please
